File: utilities/cities_import.pl
#!/usr/bin/perl -w
#
#    Procedimento para processar dados de Cidades e Estados do Brasil contidos nos arquivos : estados-cidades.js e estados-cidades2.json,
# disponibilizados no link : https://gist.github.com/letanure/3012978 e enviá-los através de POST para o sistema "adrs-restful", gravando
# assim seu conteúdo na base de dados deste sistema.

use strict;
use warnings;

use LWP;
use JSON qw(decode_json);

my $s = '{ "id": "%d", "name": "%s", "state_id": "%d" }';

my $jd = JSON->new;
my $ua = LWP::UserAgent->new;

my $endpoint = 'http://localhost:8000/city';

open(FH, '<', '/Users/daniel/Projects/GitHub/3012978/estados-cidades2.json');

my $data = do {
	local $/;
	<FH>;
};

close(FH);

my $allData = $jd->decode($data);

foreach (@{ $allData->{cities} }) {
	my $id = $_->{id};
	my $nome = $_->{name} || $_->{nome};
	my $state_id = $_->{state_id};

	print $id, "\t", $nome, "\t", $state_id, "\n";

	my $request = HTTP::Request->new(PUT => $endpoint);
	$request->header('content-type' => 'application/json');
	$request->content(sprintf($s, $id, $nome, $state_id));

	my $response = $ua->request($request);
	if ($response->is_success) {
		my $msg = $response->decoded_content;
		print "Sucesso, resposta : ", $msg, "\n";
	}
	else {
		print "Erro, resposta : ", $response->code, " -> ", $response->message, "\n";
	}

	sleep 1;
}

File: utilities/cities_import.js
#!/usr/bin/env node

"use strict";

const fs = require('fs');
const axios = require('axios');

const filename = '/Users/daniel/Projects/GitHub/3012978/estados-cidades2.json';

async function write_city(data) {
	let response = await axios.put('http://localhost:8000/city', data);
	if (response.data.data.code === 201) {
		console.log(response.data.data.data);
	}
	else {
		console.log("Error -> " + response.config.data);
	}
	/*
		.then(response => {
			if (response.data.data.code === 201) {
				console.log(response.data.data.data);
			}
		}).catch (error => {
			let not_writed = error.config.data;

			console.log('Error' + " -> " + not_writed);
		});
		*/
}

{
	let data = JSON.parse(fs.readFileSync(filename, 'utf-8'));
	let cities = data.cities;
	let kcities = Object.keys(cities);

	for (let kc = 0; kc < kcities.length; kc++) {
		let city_id = cities[kc].id;
		let city_state_id = cities[kc].state_id;
		let city_name = ( cities[kc].name || cities[kc].nome );

		let data = {
			'id': city_id,
			'name': city_name,
			'state_id': city_state_id
		};

		setTimeout(() => { write_city(data); }, 10000);
	};
}
